lib/media/selected-text-reader.ts

```diff
@@ -264,3 +310,41 @@ export async function hasSelectedText(): Promise<boolean> {
     return false
   }
 }
+
+/**
+ * Get cursor context using the new getCursorContext functionality
+ */
+export async function getCursorContext(contextLength: number): Promise<string> {
+  // find out if we have a highlighted selection already
+  const selectedText = (await getSelectedTextString(100)) || ''
+
+  const cursorContextResult = await selectedTextReaderService.getCursorContext(
+    contextLength,
+    false,
+  )
+  let preCursorText = cursorContextResult.contextText || ''
+
+  // No selected text, just get cursor context
+  if (selectedText.length === 0) {
+    return preCursorText
+  }
+
+  // If the pre-cursor text includes the selected text,
+  // we probably extended the highlighted portion, and can simply return the difference
+  // between the pre-cursor text and the selected text
+  const selectedTextIndex = preCursorText.indexOf(selectedText)
+  if (selectedTextIndex !== -1) {
+    console.log('[SelectedTextService] Selected text found in pre-cursor text')
+    preCursorText = preCursorText.slice(0, selectedTextIndex)
+    return preCursorText
+  }
+
+  // If the pre-cursor text is a subsection of selectedText, we probably deselected some text instead
+  // of getting actual pre-cursor text. Rather than trying to cut the selected text and get the
+  // pre-cursor text again, just return an empty string.
+  if (selectedText.includes(preCursorText)) {
+    return ''
+  }
```

Comment:
There are lots of edge cases, and different applications behave differently. VS Code and terminal are difficult to work with, so I tried to make this logic not disruptive to the user while getting the right context most of the time